File: mobile app/lib/services/workout_tracking_service.dart
import '../core/api_client.dart';
import '../core/api_exception.dart';
import '../models/workout_tracking_models.dart';

/// Live workout tracking — `/api/workout/*` (Strong/Hevy-style set logging).
class WorkoutTrackingService {
  WorkoutTrackingService._();
  static final WorkoutTrackingService instance = WorkoutTrackingService._();

  Future<int> resolveMemberId() async {
    try {
      final res = await ApiClient.instance.get<Map<String, dynamic>>('/workout/my-member-id');
      final id = res.data?['memberId'];
      if (id is int && id > 0) return id;
      if (id is num && id > 0) return id.toInt();
    } on ApiException catch (e) {
      if (e.statusCode != 404) rethrow;
    }
    throw const ApiException('Member profile not found. Contact your gym.');
  }

  Future<ActiveWorkoutSession> start({
    required int memberId,
    required int workoutPlanId,
  }) async {
    final offset = DateTime.now().timeZoneOffset.inMinutes;
    final res = await ApiClient.instance.post<Map<String, dynamic>>(
      '/workout/start',
      body: {
        'memberId': memberId,
        'workoutPlanId': workoutPlanId,
        'utcOffsetMinutes': offset,
      },
    );
    return ActiveWorkoutSession.fromJson(res.data ?? {});
  }

  /// Returns null when no active session (404). Supports envelope + legacy flat JSON.
  Future<ActiveWorkoutSession?> getActive(int memberId) async {
    try {
      final res = await ApiClient.instance.get<Map<String, dynamic>>(
        '/workout/active/$memberId',
      );
      final data = res.data ?? {};
      final sessionMap = data['session'] is Map
          ? Map<String, dynamic>.from(data['session'] as Map)
          : Map<String, dynamic>.from(data);
      final session = ActiveWorkoutSession.fromJson(sessionMap);
      return session.copyWith(
        lastSyncedAt: _parseDt(data['lastSyncedAt']) ?? _parseDt(sessionMap['lastSyncedAt']),
        serverTimeUtc: _parseDt(data['serverTimeUtc']) ?? _parseDt(sessionMap['serverTimeUtc']),
        pendingOfflineChanges: data['pendingOfflineChanges'] == true ||
            sessionMap['pendingOfflineChanges'] == true,
      );
    } on ApiException catch (e) {
      if (e.statusCode == 404) return null;
      rethrow;
    }
  }

  static DateTime? _parseDt(dynamic v) {
    if (v == null) return null;
    if (v is String) return DateTime.tryParse(v);
    return null;
  }

  Future<TrackedWorkoutSet> logSet({
    required int workoutSessionExerciseId,
    int? actualReps,
    double? actualWeight,
    required bool isCompleted,
    String? notes,
  }) async {
    final res = await ApiClient.instance.post<Map<String, dynamic>>(
      '/workout/log-set',
      body: {
        'workoutSessionExerciseId': workoutSessionExerciseId,
        if (actualReps != null) 'actualReps': actualReps,
        if (actualWeight != null) 'actualWeight': actualWeight,
        'isCompleted': isCompleted,
        if (notes != null && notes.isNotEmpty) 'notes': notes,
      },
    );
    return TrackedWorkoutSet.fromJson(res.data ?? {});
  }

  Future<ActiveWorkoutSession> complete(int sessionId, {double? caloriesBurned}) async {
    final res = await ApiClient.instance.post<Map<String, dynamic>>(
      '/workout/complete/$sessionId',
      query: caloriesBurned != null ? {'caloriesBurned': caloriesBurned} : null,
    );
    return ActiveWorkoutSession.fromJson(res.data ?? {});
  }

  Future<WorkoutExerciseHistory> exerciseHistory({
    required int memberId,
    required int exerciseId,
    int take = 50,
  }) async {
    final res = await ApiClient.instance.get<Map<String, dynamic>>(
      '/workout/exercise-history/$memberId/$exerciseId',
      query: {'take': take},
    );
    return WorkoutExerciseHistory.fromJson(res.data ?? {});
  }

  Future<WorkoutTrackingDashboard> dashboard(int memberId) async {
    final res = await ApiClient.instance.get<Map<String, dynamic>>(
      '/workout/dashboard/$memberId',
    );
    return WorkoutTrackingDashboard.fromJson(res.data ?? {});
  }
}

File: gym_client/src/services/workoutTracking.service.ts
import { api } from '../lib/api'
import type {
  ActiveWorkoutActiveResponse,
  ActiveWorkoutSession,
  MemberWorkoutSummary,
  MemberWorkoutTimeline,
  WorkoutAdminMonitoring,
  WorkoutDashboard,
  WorkoutExerciseHistory,
  WorkoutSessionDetail,
  WorkoutSessionExercise,
} from '../types/workoutTracking'

const base = '/workout'

export const workoutTrackingService = {
  myMemberId: () => api.get<{ memberId: number }>(`${base}/my-member-id`),

  start: (body: { memberId: number; workoutPlanId: number; utcOffsetMinutes?: number }) =>
    api.post<ActiveWorkoutSession>(`${base}/start`, body),

  getActive: (memberId: number) => api.get<ActiveWorkoutActiveResponse>(`${base}/active/${memberId}`),

  getActiveSession: async (memberId: number) => {
    const { data } = await api.get<ActiveWorkoutActiveResponse>(`${base}/active/${memberId}`)
    return data.session
  },

  logSet: (body: {
    workoutSessionExerciseId: number
    actualReps?: number
    actualWeight?: number
    durationSeconds?: number
    restSeconds?: number
    isCompleted: boolean
    notes?: string
  }) => api.post<WorkoutSessionExercise>(`${base}/log-set`, body),

  complete: (sessionId: number, caloriesBurned?: number) =>
    api.post<ActiveWorkoutSession>(`${base}/complete/${sessionId}`, null, {
      params: caloriesBurned != null ? { caloriesBurned } : undefined,
    }),

  exerciseHistory: (memberId: number, exerciseId: number, take = 50) =>
    api.get<WorkoutExerciseHistory>(`${base}/exercise-history/${memberId}/${exerciseId}`, {
      params: { take },
    }),

  dashboard: (memberId: number) => api.get<WorkoutDashboard>(`${base}/dashboard/${memberId}`),

  trainerMemberWorkouts: (take = 30) =>
    api.get<MemberWorkoutSummary[]>(`${base}/trainer/members`, { params: { take } }),

  trainerMemberTimeline: (memberId: number, take = 40) =>
    api.get<MemberWorkoutTimeline>(`${base}/trainer/members/${memberId}/timeline`, { params: { take } }),

  sessionDetail: (sessionId: number) =>
    api.get<WorkoutSessionDetail>(`${base}/session/${sessionId}/detail`),

  adminMonitoring: (take = 50) =>
    api.get<WorkoutAdminMonitoring>(`${base}/admin/monitoring`, { params: { take } }),
}
